CASSANDRA-17111: Minimal Deps

belliottsmith opened the pull request.

dcapwell commented on accord-core/src/main/java/accord/api/Key.java:
```diff
@@ -3,10 +3,7 @@
-     * Returns a hash code of a key to support accord internal sharding. Hash values for equal keys must be equal.
-     */
-    int keyHash();
+    Key toRoutingKey();
```
`Key` is generic, should this be `K` instead?  Also what is the reasoning behind moving `keyHash` to a new `RoutingKey` and adding a method to return that type?  

As far as I can see `toRoutingKey` is dead code?  I don't find anything that calls it

belliottsmith replied:
In the next patch set I remove the generic parameter, as it just pollutes the codebase to no benefit, as we never specify it. It simplifies only one method at the expense of lots of other code.

`toRoutingKey` appears to be dead in this patch set yes, I'll remove it.

`RoutingKey` overall is used in the later patch set in part to minimise the data that is sent, and in part to better formalise the relationship keys and their use cases (for routing, we only need a token, not a full DecoratedKey in C* terms)

dcapwell commented on accord-core/src/main/java/accord/messages/Callback.java:
```diff
@@ -9,5 +9,7 @@
-    void onFailure(Id from, Throwable throwable);
+    default void onSlowResponse(Id from) {}
+    void onFailure(Id from, Throwable failure);
+    void onCallbackFailure(Throwable failure);
```
can we add `Id from` to be consistent?  Every call path I see knows `from`.

As far as I can see the only places that call this are tests, and they have a different pattern than the rest of the code...

src/main

```
callback.onSuccess(from, response);
```

src/test

```
try
{
  callback.onSuccess(from, response);
}
catch (Throwable t)
{
  callback.onCallbackFailure(t);
}
```

What is the reason for the divergence?

belliottsmith replied:
> can we add Id from to be consistent? Every call path I see knows from.

I'm not sure it's as helpful here, as we're talking about internal logic failures? Though I suppose there's no harm in logging the source of the message that caused a problem.

> What is the reason for the divergence?

I take it you're referring to the maelstrom implementation? You're right though, these should be updated as well.

dcapwell commented on accord-core/src/main/java/accord/coordinate/Execute.java:
```diff
@@ -1,88 +1,78 @@
-import org.apache.cassandra.utils.concurrent.AsyncPromise;
-import org.apache.cassandra.utils.concurrent.Future;
 
-class Execute extends AsyncPromise<Result> implements Callback<ReadReply>
+class Execute implements Callback<ReadReply>
```
Why move away from Promise/Future in favor of Callback?  With futures we know that setting can only happen once, but now we need to worry about calling callback multiple times or hoping the callback maps to a future to deal with that.

belliottsmith replied:
I am a strong believer in using minimal functionality. Futures are actually pretty heavy-weight and complex objects (that also heavily pollute call-graphs and control flow). Here we are passing a single callback through the entire hierarchy of commands, with each one either invoking it (and terminating) or passing it on to another for responsibility. We save memory, CPU time, and it is also cleaner (IMO).

We could plausibly pass a `Future` through the chain to be completed, but we're still ultimately registering precisely one callback, so we're just introducing an entirely unnecessary intermediate object.

If you would like to, I would be OK with introducing an `AtMostOnceCallback` object, that guards Callbacks to ensure they're invoked at most once, as this would be relatively lightweight, and keeps the intent and control flow pretty much as simple and straightforward. But I think it is fine either way.

dcapwell commented on accord-core/src/main/java/accord/api/Query.java:
```diff
@@ -9,5 +9,5 @@
      * from some {@link DataStore}, to produce a {@link Result} to return to the client.
      */
-    Result compute(Data data);
+    Result compute(Data data, Read read, Update update);
```
should `Update` have the `Nullable` annotation?  It is nullable on read-only queries

belliottsmith replied:
Yep, happy to add that thanks.

dcapwell reviewed (commented):
trying to break this up into smaller pieces, so trying to look at the simpler stuff before the core changes

dcapwell commented on accord-core/src/main/java/accord/local/CommandStores.java:
```diff
@@ -208,10 +207,10 @@ KeyRanges currentRanges()
         }
 
-        private static long addKeyIndex(Key key, long numShards, long accumulate)
+        private static long addKeyIndex(int i, Key key, long numShards, long accumulate)
```
`i` is dead code, is this still needed?

belliottsmith replied:
I think it's probably to permit its use as a method handle, perhaps we should have an annotation or shorthand for that.

dcapwell replied:
yeah, it looks like that is the case, to still allow would need a 

```
public long foldl(KeyRanges rs, FoldToLong<Key> fold, long param, long initialValue, long terminalValue)
```

but that adds more code... so this is fine.

dcapwell commented on accord-core/src/main/java/accord/messages/BeginRecovery.java:
```diff
@@ -78,17 +78,19 @@ public void process(Node node, Id replyToNode, ReplyContext replyContext)
                 earlierCommittedWitness = committedStartedBefore(instance, txnId, txn.keys)
                                           .filter(c -> c.savedDeps().contains(txnId))
-                                          .collect(Dependencies::new, Dependencies::add, Dependencies::addAll);
+                                          .collect(() -> Deps.builder(txn.keys), Deps.Builder::add, (a, b) -> { throw new IllegalStateException(); })
```
same logic as line 89, maybe best if we add a `Deps.collector(txn.keys)` that looks like the following

```
public static Collector<Command, Builder, Builder> collector(Keys keys)
{
    return Collector.of(() -> Deps.builder(keys), Deps.Builder::add, (a, b) -> { throw new IllegalStateException(); });
}
```

dcapwell commented on accord-core/src/main/java/accord/txn/Txn.java:
```diff
@@ -4,12 +4,15 @@
 
 public class Txn
 {
-    enum Kind { READ, WRITE, RECONFIGURE }
```
removing `RECONFIGURE` for now or is this now handled differently?  Want to make sure I don't miss.

belliottsmith replied:
It was introduced in a very early iteration and never ended up making sense, I had just failed to notice it hadn't been removed.

dcapwell commented on accord-core/src/main/java/accord/utils/Timestamped.java:
```diff
@@ -17,4 +17,10 @@ public static <T> Timestamped<T> merge(Timestamped<T> a, Timestamped<T> b)
     }
+
+    @Override
+    public String toString()
```
given that this also has timestamp would it make sense to add to string as well?

belliottsmith replied:
Maybe, though I think when debugging it's important to balance information, and it's easier for humans to parse maps with just the value string IME. It's rare the timestamp is important, so I preferred to prioritise the more common case.

dcapwell commented on accord-core/src/main/java/accord/api/Key.java:
```diff
@@ -3,10 +3,7 @@
-     * Returns a hash code of a key to support accord internal sharding. Hash values for equal keys must be equal.
-     */
-    int keyHash();
+    Key toRoutingKey();
```
> toRoutingKey appears to be dead in this patch set yes, I'll remove it.

just reposting so it doesn't get lost, seems after the last update It doesn't show up while reading the code.

belliottsmith replied:
Last update was just a rebase to get rid of the conflicts, I haven't addressed your comments yet - sorry for any confusion!

dcapwell replied:
no confusion at all!  I personally rely on GH comments to make sure nothing gets dropped, so when GH starts to hide things I like to re-comment just to make sure it doesn't get lost.

belliottsmith replied:
I do the same, I just rely on the 'resolved' indicator, and make sure to expand the 'Outdated' ones

dcapwell reviewed (commented):
Ok did my next path, looked through everything simple that doesn't depend on changes to Deps or new InlineHeap/SortedArrays; I want to make sure behavior is the same, so will take a bit to confirm that.

Tomorrow ill try to go through `Deps`... its a lot so will need to do this itself in isolation.

belliottsmith commented:
I do the same, I just rely on the 'resolved' indicator, and make sure to expand the 'Outdated' ones

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+            int txnIdx = ensureTxnIdx(txnId);
+            int keyIdx = keys.indexOf(key);
+            if (keysToTxnIdCounts[keyIdx] == keysToTxnId[keyIdx].length)
+                keysToTxnId[keyIdx] = Arrays.copyOf(keysToTxnId[keyIdx], Math.max(4, keysToTxnIdCounts[keyIdx] * 2));
```
don't think we need `Math.max(4`, it should already start with 4 and we don't have the same logic in `add(Command command)`

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+    public static class Builder
+    {
+        final Keys keys;
+        final Map<TxnId, Integer> txnIdLookup = new HashMap<>(); // TODO: primitive map
```
Should we just add HPPC (https://github.com/carrotsearch/hppc)?  Its already in C* and could simplify this class a little bit.

belliottsmith replied:
Sure. I wish we had better primitive maps available to us, but HPPC will certainly do.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+        private int txnIdx(TxnId txnId)
+        {
+            return txnIdLookup.getOrDefault(txnId, -1);
+        }
```
We could remove the `txnIdx` function and just have `contains` do `return txnIdLookup.containsKey(txnId);`

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+
+        public boolean isEmpty()
+        {
+            return Arrays.stream(keysToTxnIdCounts).allMatch(i -> i == 0);
```
maybe just `return txnIdLookup.isEmpty()`?  I point out in a different comment the code currently allows `!txnIdLookup.isEmpty()` but `keysToTxnIdCounts` is empty, so guess if we want to keep that behavior then this logic is more correct to me.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+        public void add(Command command)
+        {
+            int idx = ensureTxnIdx(command.txnId());
+            keys.foldlIntersect(command.txn().keys, (li, ri, k, p, v) -> {
```
if higher up logic is incorrect, we could see a `Command` trying to be added that doesn't have any conflicting keys, this would lead to `TxnId` being updated with no keys being touched... Should we detect this and fail?  `add(Key key, TxnId txnId)` will fail due to trying to insert into an array at a negative offset.

dcapwell reviewed (commented):
Focused on `Deps.Builder` mostly trying to make sure I fully grasp each line.  I created a review branch (https://github.com/dcapwell/cassandra-accord/tree/review/minimal-deps) and added a commit trying to improve docs https://github.com/dcapwell/cassandra-accord/commit/368ce26374722994ab0a1d80820b61254f540b13. Anything that took a bit to reason about I tried to add docs after doing so.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+        }
+    }
+
+    public boolean checkValid()
```
Pushed a commit https://github.com/dcapwell/cassandra-accord/commit/3d17394952e2d498b190ab544c997a7fe9de3fbe.

I feel that this method should be what throws as it could produce a much more meaningful error than the caller can.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+
+    public boolean equals(Deps that)
+    {
+        return this.txnIds.length == that.txnIds.length
```
`this.txnIds.length == that.txnIds.length` is checked in `Arrays.equals(this.txnIds, that.txnIds)`

belliottsmith replied:
The idea here is to abort quickly with size checks first, before we perform any array comparisons that might succeed where we might have avoided them by one of the other size checks.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+    }
+
+    @Override
+    public boolean equals(Object o)
```
Override `equals` without `hashCode`, can we add one?

```
@Override
    public int hashCode() {
        int result = Objects.hash(keys);
        result = 31 * result + Arrays.hashCode(txnIds);
        result = 31 * result + Arrays.hashCode(keyToTxnId);
        result = 31 * result + Arrays.hashCode(txnIdToKey);
        return result;
    }
```

belliottsmith replied:
I disagree with this policy / linter advice. If an object is not being used as a key in a hash map, and is not intended to be, it should not be given a `hashCode()` method (and indeed, the Cassandra style guide now stipulates this, though I'm sure this commonly followed policy will win out regularly)

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+        }, null);
+    }
+
+    public void forEachOn(CommandStore commandStore, Timestamp executeAt, Consumer<TxnId> forEach)
```
This is used in one place, so kinda feel it would be best to drop this public method and just have the one caller do this logic, this also means `Deps` doesn't depend on `CommandStore`

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+    }
+
+    @Inline
+    public long foldl(KeyRanges rs, IndexedFoldToLong<Key> fold, long param, long initialValue, long terminalValue)
```
feel we can get rid of `long param` as its just there to pass a `shards.length` to `ShardedRanges::addKeyIndex`, every other call path has access to it already.

belliottsmith replied:
If we get rid of it then `ShardedRanges::addKeyIndex` likely becomes a non-static lambda.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+                }
+
+                return value;
+            }, offset, 0, 0xffffffffffffffffL);
```
I think `-1L` is best,  with `0xffffffffffffffffL` I have to count all the `f` to know how many bits are set

belliottsmith replied:
I used to get complaints that -1 wasn't clear enough, which is why I switched to 0xf.... I've always preferred -1. Happy to switch.

dcapwell reviewed (commented):
Pushed more review commits, mostly adding tests or docs to the code.  I am most of the way through `Deps`, just have `with` and `merge` left.

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+
+    /**
+     * Given two sorted arrays, where one is a subset of the other, return an int[] of the same size as
+     * the {@code src} parameter, with the index within {@code trg} of the corresponding element within {@code src},
```
its not the same size as the `src`, its `srcLength`.  We also don't validate that `srcLength` is within bounds, so a mess-up would lead to an out of bounds exception.  This method is also not directly used, the other `remapper` is the only one used.

Might be best to move the doc to the other one and remove the `length` params?

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+    public static <T extends Comparable<? super T>> int[] remapper(T[] src, int srcLength, T[] trg, int trgLength, boolean trgIsSuperset)
+    {
+        if (src == trg || (trgIsSuperset && trgLength == srcLength)) return null;
+        int[] result = new int[srcLength];
```
needs a `Arrays.fill(result, -1);` else we put `0` as the case where trg ends *before* src, in my PR I do this and created property tests to show this.

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+     * Given two sorted arrays, where one is a subset of the other, return an int[] of the same size as
+     * the {@code src} parameter, with the index within {@code trg} of the corresponding element within {@code src},
+     * or -1 otherwise.
+     * That is, {@code src[i] == -1 || src[i].equals(trg[result[i]])}
```
`result[I] == -1`, my PR adds this

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+    }
+
+    @Inline
+    public static int exponentialSearch(int[] in, int from, int to, int find)
```
no code touches this

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+     * Copy-on-write insert into the provided array; returns the same array if item already present, or a new array
+     * with the item in the correct position if not. Linear time complexity.
+     */
+    public static <T extends Comparable<? super T>> T[] insert(T[] src, T item, IntFunction<T[]> factory)
```
nothing references

belliottsmith replied:
Might be used in follow up PR. I’ll double check, but this PR is in part designed as a split to separate the non-protocol changes that are dependencies for the follow on patch.

belliottsmith replied:
It is used in partial-replication

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+        }
+    }
+
+    public static <T> Deps merge(Keys keys, List<T> merge, Function<T, Deps> getter)
```
I was curious why have a custom merge method rather than rely on builder, so wanted to create a quick benchmark so I could see if there was any noticeable difference, but in doing so I find that merge produces an invalidate Deps for me

the simplified test (was benchmark, so had other non merge related things)

```
private static accord.primitives.Deps benchmarkMerge(List<Deps> list) {
        Keys keys = list.get(0).test.keys();
        return accord.primitives.Deps.merge(keys, list, d -> d.test);
    }
```

The error

```
Seed=3706255249468386534
Examples=100
accord.utils.Property$PropertyError: Seed=3706255249468386534
Examples=100
	at accord.utils.Property$SingleBuilder.check(Property.java:78)
	at accord.txn.DepsTest.hackyBenchmark(DepsTest.java:186)
...
Caused by: java.lang.IllegalStateException
	at com.google.common.base.Preconditions.checkState(Preconditions.java:494)
	at accord.primitives.Deps.<init>(Deps.java:377)
	at accord.primitives.Deps.merge(Deps.java:345)
	at accord.txn.DepsTest.benchmarkMerge(DepsTest.java:231)
	at accord.txn.DepsTest.lambda$hackyBenchmark$15(DepsTest.java:188)
	at accord.utils.Property$SingleBuilder.check(Property.java:74)
	... 91 more
```

Was trying to test before I started reviewing as `merge` is very large, so not sure where the issue is yet

dcapwell replied:
Switching `Keys` to be 

```
Keys keys = list.stream().map(d -> d.test.keys()).reduce(Keys.EMPTY, (l, r) -> Keys.union(l, r));
```

gets `merge` to no longer fail, so was able to get my hacked up benchmarks (pushed to review/minimal-deps but should remove after review is done)

```
Benchmark[name=merge(size=6, keys=279)] avg 131.05ms, allocated 5357 MiB
Benchmark[name=builder forEach(size=6, keys=279)] avg 398.75ms, allocated 9418 MiB
Benchmark[name=merge(size=14, keys=300)] avg 555.55ms, allocated 16332 MiB
Benchmark[name=builder forEach(size=14, keys=300)] avg 898.02ms, allocated 18628 MiB
Benchmark[name=merge(size=11, keys=320)] avg 343.21ms, allocated 11180 MiB
Benchmark[name=builder forEach(size=11, keys=320)] avg 736.06ms, allocated 15173 MiB
Benchmark[name=merge(size=15, keys=342)] avg 508.55ms, allocated 16129 MiB
Benchmark[name=builder forEach(size=15, keys=342)] avg 919.88ms, allocated 19336 MiB
Benchmark[name=merge(size=16, keys=355)] avg 689.76ms, allocated 22904 MiB
Benchmark[name=builder forEach(size=16, keys=355)] avg 1,162.81ms, allocated 22445 MiB
```

I do see that if you are in a loop merging `Deps` non stop that `merge` is faster but the amount of allocated memory isn't that much smaller than the `Builder` way

belliottsmith replied:
merge currently has a handful of trivially cacheable large allocations, so it will be much faster and lower allocation when those simple TODO are completed. It was more important to me to do the difficult work first, but you’re welcome to tackle the TODOs as part of review to make it faster if you like! (Probably easier once we have imported c* utils)

dcapwell replied:
> welcome to tackle the TODOs as part of review to make it faster if you like

Right now I am focused on correctness rather than performance, so was hoping that there wasn't a clear win so I could argue to remove in favor of simpler code.  I made an op in builder to allocate less, so now `builder` allocates less than `merge`, but `merge` is still ~30% faster.

Just because I already had everything in place, here is a flame graph of both code paths:

merge
<img width="1709" alt="Screen Shot 2022-08-09 at 8 50 57 AM" src="https://user-images.githubusercontent.com/1852601/183699251-8004e511-e518-43f7-a8eb-0afbf48d5ff7.png">

builder
<img width="1722" alt="Screen Shot 2022-08-09 at 8 51 21 AM" src="https://user-images.githubusercontent.com/1852601/183699355-df064618-3781-41d1-8925-694a2ee0c13c.png">

belliottsmith replied:
Merge will allocate close to nothing when it uses caching, there is zero chance builder will be able to allocate less.

belliottsmith replied:
Looking at the flame graphs I expect we can make merge faster too, although the behaviour is of course highly dependent on the data you're generating.

dcapwell replied:
yep, my tests bias towards large number of keys (200+) but when smaller there really isn't a difference between the two.

Just pushed a property test to validate merge and builder "do the right thing", and this shows that merge currently doesn't work with partial keys; trying to figure out why now.

belliottsmith replied:
Wdym partial keys?

The number of deps we are merging, the number of txns per key, and the amount of overlap between the deps  and distribution of that overlap all matter a great deal. Again, a performance comparison should also only really be made with the simple caching implemented (for a test this could be as simple as a thread local of the maximum allocated size of each buffer)

belliottsmith replied:
Oh, gosh, there's a huge amount of performance left on the table. We're not efficiently merging `TxnId` (the TODO depends on a cassandra-utils jar or copy/pasting `MergeIterator`), let alone optimisations to `remapper`.

So far the only important work done is optimally merging the `int[]`, none of the other setup (which is why it is only 1/3rd of the time at present)

I assume your test generates entirely random keys and `TxnId`? Do you have a working branch? I'll take a look either later or tomorrow.

belliottsmith replied:
> Wdym partial keys?

So, it looks like you mean if you were to provide a `Keys` object that does not contain all of the keys of the lists you are merging. In this case, the method is not expected to succeed. We can document that, or if you prefer I can modify it to remove data for the Keys that are missing.

dcapwell replied:
> In this case, the method is not expected to succeed

the usage of `merge` looks like the following at all call paths

```
Deps.merge(txn.keys, recoverOks, ok -> ok.earlierAcceptedNoWitness)
```

so always taking the `oks.deps` and merging.  Based off what I think you are saying, the `txn.keys` has a hard requirement that it contains every key found in `oks.deps.keys`?  If so a bug in the `oks` (such as we forget to strip out non intersecting keys) logic would then cause the coordinator to fail during `merge` rather than remove any unreferenced keys?  So the desired behavior in this case is not to "merge" the Deps but instead to reject the input?

dcapwell replied:
> a performance comparison should also only really be made with the simple caching implemented

Sorry for causing any confusion for motivations behind the perf tests I hacked together, this was to help me try to understand the motivations behind the code.  The code is dense and not documented, so "what should it do" is something I infer based off the name, so I saw another way to handle with `Builder` so I wanted to ask "what is the gain we see"; this work is meant to be thrown away and was to help me justify the time to review rather than replace.

belliottsmith replied:
> If so a bug in the oks (such as we forget to strip out non intersecting keys) logic would then cause the coordinator to fail during merge rather than remove any unreferenced keys?

If there's a bug in the `ok` response all bets are off for correctness, so yeah we probably want to fail if it somehow includes an unexpected key (and in fact if you like we can optionally enforce this in `remapper`, though if we want to detect this it would be better to do so in the callback so we do not fail the entire operation)

dcapwell replied:
cool, will correct the test and try to have this detected and produce a cleaner error

dcapwell replied:
pushed a change to 574bcd14cfce7073787ea430e0db79b01bad76b7 to detect this in remapper when targetSuperset=true, I figured I will keep the old behavior when its false; all call paths are true today

belliottsmith replied:
> my tests bias towards large number of keys (200+) but when smaller there really isn't a difference between the two.

Interesting. I just checked out your PR and made some tweaks to improve `remapper`, use `MergeIterator` and cache `int[]`, and while the performance is relatively stable for large numbers of keys (I haven't compared for small numbers of keys), I see a performance delta that is much larger between the two approaches for smaller numbers of keys.

```
13:22:43.148 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=2, keys=7, size=904)] avg 7.65ms, 0.21ms GC, allocated 594 MiB
13:22:45.846 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=2, keys=7, size=904)] avg 63.79ms, 0.22ms GC, allocated 3289 MiB
13:22:45.872 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=1, keys=14, size=149)] avg 0.06ms, 0.17ms GC, allocated 4 MiB
13:22:46.145 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=1, keys=14, size=149)] avg 6.13ms, 0.21ms GC, allocated 340 MiB
13:22:57.425 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=11, keys=21, size=3863)] avg 249.98ms, 0.29ms GC, allocated 3297 MiB
13:23:12.559 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=11, keys=21, size=3863)] avg 359.93ms, 0.23ms GC, allocated 11533 MiB
13:23:15.202 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=3, keys=13, size=1512)] avg 56.32ms, 0.23ms GC, allocated 1400 MiB
13:23:21.237 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=3, keys=13, size=1512)] avg 143.27ms, 0.22ms GC, allocated 5531 MiB
```

belliottsmith replied:
For comparison, I included running `with` in a loop, and to my surprise this is often faster. It allocates a lot more garbage, but this could readily be optimised by separating `with` into a function that operates on provided in/out arrays, and re-using pairs of buffers.

Given `merge` has lower complexity than `with` it suggests we're leaving some constant factors on the table. Since it's faster even with only 4 streams to merge, it's not as simple as cache associativity, however optimally we might anyway prefer to buffer the input streams in a cache-sympathetic manner. It's possible that a sequence of linear merges has better pipelining / predictability.

It's worth noting that in the real world, we expect to mostly merge either disjoint keys or keys with `txnId` lists that are mostly the same. So, we should run some tests specifically with this in mind to verify approaches. We don't anticipate lots of keys, but probably 9-16 copies for each key, and dozens to hundreds of dependencies per transaction.

Either way, perhaps it would be fine to simplify `merge` into a variant of recursive `with` that produces less garbage, and later perhaps introduce a variant that balances algorithmic complexity (which is very poor with this approach) and constant factors.


```
13:47:29.482 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=4, keys=21, size=1646)] avg 168.35ms, 0.22ms GC, allocated 6901 MiB
13:47:31.980 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=4, keys=21, size=1646)] avg 59.03ms, 0.24ms GC, allocated 1763 MiB
13:47:33.485 [main] INFO accord.txn.DepsTest - Benchmark[name=with(No.=4, keys=21, size=1646)] avg 35.4ms, 0.22ms GC, allocated 2308 MiB
13:48:15.296 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=15, keys=27, size=8449)] avg 951.11ms, 0.23ms GC, allocated 30617 MiB
13:48:48.707 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=15, keys=27, size=8449)] avg 795.14ms, 0.22ms GC, allocated 13242 MiB
13:49:16.317 [main] INFO accord.txn.DepsTest - Benchmark[name=with(No.=15, keys=27, size=8449)] avg 656.92ms, 0.23ms GC, allocated 35741 MiB
13:49:44.247 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(No.=9, keys=22, size=5812)] avg 638.99ms, 0.24ms GC, allocated 22966 MiB
13:50:01.518 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(No.=9, keys=22, size=5812)] avg 410.77ms, 0.23ms GC, allocated 8037 MiB
13:50:14.952 [main] INFO accord.txn.DepsTest - Benchmark[name=with(No.=9, keys=22, size=5812)] avg 319.43ms, 0.23ms GC, allocated 16794 MiB
```

dcapwell replied:
> It's worth noting that in the real world, we expect to mostly merge either disjoint keys or keys with txnId lists that are mostly the same. So, we should run some tests specifically with this in mind to verify approaches. We don't anticipate lots of keys, but probably 9-16 copies for each key, and dozens to hundreds of dependencies per transaction.

This makes a lot of sense, if you happen to touch the same key vs you are the same txn just with slightly different inputs so all keys conflict.

> but probably 9-16 copies for each key, and dozens to hundreds of dependencies per transaction.

yeah, also makes sense, use smaller number of keys with larger `Deps` list.

belliottsmith replied:
> This makes a lot of sense, if you happen to touch the same key vs you are the same txn just with slightly different inputs so all keys conflict.

More that a given transaction is _likely_ to arrive at all replicas at around the same "point" (i.e. in the same order wrt other transactions) - and with our 1RT optimisation is all but _guaranteed_ to do so (excluding lost messages to some replicas), each key is likely to have _precisely_ the same txnIds _most of the time_, and sometimes just vary a little. But each shard will cover only some (different for each replica) subset of the keys, so we're going to normally be copying over the txnIds for each key and validating they're the same.

Now that I say this, it makes potentially more sense to perform the efficient `with` variation, even if it is algorithmically worse, as most of the time it will be very efficient. Given that we _aren't_ testing this kind of input and it still performs adequately, the simplicity is probably well justified.

dcapwell replied:
> perform the efficient with variation ... the simplicity is probably well justified.

WFM, ill focus on `with` and `without` and ignore `merge` for now

dcapwell reviewed (commented):
Didn't make as much progress as I hoped; prod issue took most of my day.

I was trying to review `Deps.merge` but that depends on SortedArrays so tried to review and test that part of the code, so will try to get to `Deps.merge` tomorrow

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+            }
+
+            if (result == null)
+                return left;
```
why return left in this case?  Been working on tests for this class and this fails in the following case

```
Property error detected:
Seed = 3576953691942488024
Examples = 1000000
Pure = true
Error: array lengths differ, expected: <0> but was: <1>
Values:
	0 = [1459905072]
	1 = [-1799119784, -1117722578, -1072178078, -1012124199, -804935562, -688955866, 95532926, 410144613, 604870749, 1070710384]
```

test

```
@Test
    public void testLinearIntersection()
    {
        Gen<Integer[]> gen = Gens.arrays(Integer.class, Gens.ints().all())
                .unique()
                .ofSizeBetween(0, 100)
                .map(a -> {
                    Arrays.sort(a);
                    return a;
                });
        qt().withSeed(3576953691942488024L).forAll(gen, gen).check((a, b) -> {
            Set<Integer> left = new HashSet<>(Arrays.asList(a));
            Set<Integer> right = new HashSet<>(Arrays.asList(b));
            Set<Integer> intersection = Sets.intersection(left, right);
            Integer[] expected = intersection.toArray(Integer[]::new);
            Arrays.sort(expected);

            Integer[] actual = SortedArrays.linearIntersection(a, b, Integer[]::new);
            Assertions.assertArrayEquals(expected, actual);
        });
    }
```

So, when we see there isn't an overlap we return `left`  when left is smaller than right, else we return `right`?  This seems confusing to me as the caller can never detect the case where there is zero intersections

belliottsmith replied:
Quite likely a bad copy/paste/edit from `linearUnion`, that I guess hasn't had any impact yet.. Good catch

dcapwell replied:
pushed patch to detect this and return empty array, see 3665046941c4499ad4e7d975295fdcfe1686c34b

dcapwell replied:
added special test to validate the optimization where smallest is a perfect subset of the larger array (`larger[x:y]`)

dcapwell replied:
added 9b6e0225a6883d16459d0030952af78a08929349, my patch didn't check `result == null` so when RHS was larger this caused a regression..

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+            }
+
+            if (result == null)
+                return right;
```
same comment as above, why return `right` when we don't have any intersections?

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+            }
+        }
+
+        if (resultSize < result.length)
```
doesn't properly handle the case where left is larger than right, need the below to handle it

```
while (leftIdx < left.length)
            result[resultSize++] = left[leftIdx++];
```

Fixed in 42e93d1686fe4eb1d7972505c09f64c7f463c181

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+    }
+
+    // TODO: check inlining elides this
+    public enum Search { FLOOR, CEIL, FAST }
```
Spoke in slack, can we add docs for this?  This is similar to `java.util.NavigableMap#floorEntry` and `java.util.NavigableMap#ceilingEntry` but would be good if the reader can read the docs rather than the code usage.

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRanges.java:
```diff
@@ -0,0 +1,395 @@
+    public KeyRanges(KeyRange[] ranges)
+    {
+        Preconditions.checkNotNull(ranges);
+        this.ranges = ranges;
```
`Keys.slice` has a requirement that these are sorted, yet I don't see any code that enforces this... the only caller I see that actually builds ranges is `accord.topology.Topology#Topology(long, accord.topology.Shard...)` but that is the user... so should this sort the ranges? And if so how? 

Just pushed tests for `Keys.slice` and hitting the fact that overlapping ranges tricks it, so it starts to filter out even though the ranges include all Ints

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+
+    public Keys slice(KeyRanges ranges)
+    {
+        return wrap(SortedArrays.sliceWithOverlaps(keys, ranges.ranges, Key[]::new, (k, r) -> -r.compareTo(k), KeyRange::compareTo));
```
in trying to review `SortedArrays.sliceWithOverlaps` I figured it would be easier to test this method rather than add tests for `SortedArrays.sliceWithOverlaps` directly, but got bitten by the fact that nothing forces `KeyRanges` to be sorted, and overlapping ranges don't have a clear sort order.

belliottsmith replied:
Perhaps the method is poorly named, but the overlaps refers to the fact that a range may match multiple other ranges or keys on the other side of the slice, not that either side may permit overlaps. KeyRanges self enforces sorting and no overlaps for all of its own actions, so only the external inputs need to be cleaned, if any. I will add javadoc to explain all of this.

dcapwell replied:
> KeyRanges self enforces sorting and no overlaps for all of its own actions

Where do we `enforce`?  Most methods make an assumption that the data is sorted and without overlaps, but nothing sorts this or fails on overlaps, and the only caller in `src` is topology, which uses the order from the user (in our case C*)... so I have yet to find anything that makes this a sorted array or disallow range overlaps.

> Perhaps the method is poorly named ... not that either side may permit overlaps

I didn't do range overlaps because this called itself `overlaps`, I did range overlaps because `KeyRanges` "allows" it as it never sanitizes its inputs (if inputs were 100% controlled then I wouldn't do this, but the inputs are from users that provide `Shard`, so the correctness of this class is 100% dependent on users).  Since I couldn't find any ordering or control of its inputs, I went with what it "allows" regardless of what its methods assume, this then drives the statement that I posted, that what is allowed vs what is expected doesn't match.

> if any. I will add javadoc to explain all of this.

In this case I don't feel a javadoc is best, the input to this is from users, so without sanitizing to enforce assumptions, this isn't really safe.  Also looking at the single usage (creation) its done once per topology, so validating assumptions is something we could do on that path and then all the internal methods can avoid those checks as the assumptions were already handled.

belliottsmith replied:
> validating assumptions is something we could do on that path

Yes, this what I meant by

> so only the external inputs need to be cleaned, if any

belliottsmith replied:
To be clear, my intention/preference was to make the constructor private and publish a static constructor method that sorts and de-overlaps.

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,535 @@
+                resultCount = ai;
+                result = factory.apply(ai + (slice.length - nextai));
+                System.arraycopy(slice, 0, result, 0, resultCount);
+                ai = nextai;
```
why don't we update `ri`?  I am in a debugger trying to see the reason and not seeing a clear reason, setting `ri = (int)ari;` is passing in my tests.

The side effect I see by not setting it is that 

```
            int nextai = exponentialSearch(slice, ai, slice.length, select[ri], cmp2, Search.FLOOR) + 1;
             while (ai < nextai)
                 result[resultCount++] = slice[ai++];
```

is always a wasted operation on the first call, this is because `ri` wasn't updated so we *know* a new `ri` is the next match so we search for a match that doesn't exist.

dcapwell replied:
I added it in cd51a71795d48010be07aab7c583ac9bfd19913f with docs for this method

belliottsmith replied:
Yes, I think you're right. I'll think about it more carefully with fresh eyes in the morning.

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+    }
+
+    // TODO: optimise for case where none removed
+    public Deps without(Predicate<TxnId> remove)
```
ignoring `merge` method for now, so this is the last one for this review!  I pushed tests and it is failing for me, trying to figure out why but what I see is the following: remove T1 (which exists) and check key -> [T] to see if still present, and check someKeys(T) -> [K] as well... what I am seeing is that a different T2 gets removed instead....  but T2 is known in the txnId array, its just that the offsets are messed up for this txn only

```
accord.primitives.Deps without = deps.test.without(i -> i.equals(txnId));
```

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,978 @@
+                keyToTxnId[k++] = o;
+
+            int remapped = remapTxnIds[this.keyToTxnId[i]];
+            if (remapped > 0)
```
found the issue, needs to be `>=`

belliottsmith commented:
I've pushed (a very in progress) update to my branch of your review branch, and I'm seeing the following:

```
00:29:49.463 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=7, keys=257, total=3400)] avg 134.95ms, allocated 1907 MiB
00:30:07.839 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=7, keys=257, total=3400)] avg 427.51ms, allocated 13019 MiB
00:30:46.528 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=20, keys=344, total=11182)] avg 857.73ms, allocated 20997 MiB
00:31:43.433 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=20, keys=344, total=11182)] avg 1,344.11ms, allocated 37105 MiB
00:32:02.101 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=12, keys=326, total=6410)] avg 404.94ms, allocated 5409 MiB
00:32:35.870 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=12, keys=326, total=6410)] avg 795.45ms, allocated 24247 MiB
00:32:54.877 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=12, keys=355, total=7099)] avg 410.93ms, allocated 5316 MiB
00:33:31.908 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=12, keys=355, total=7099)] avg 870.99ms, allocated 25560 MiB
00:33:38.520 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=8, keys=256, total=2991)] avg 134.86ms, allocated 2062 MiB
00:33:55.547 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=8, keys=256, total=2991)] avg 395.47ms, allocated 12105 MiB
00:33:56.510 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=2, keys=90, total=1536)] avg 9.84ms, allocated 516 MiB
00:34:03.143 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=2, keys=90, total=1536)] avg 149.17ms, allocated 5743 MiB
00:34:35.287 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=20, keys=337, total=7961)] avg 711.37ms, allocated 10672 MiB
00:35:20.816 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=20, keys=337, total=7961)] avg 1,073.09ms, allocated 28380 MiB
00:35:28.808 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=8, keys=308, total=3777)] avg 164.52ms, allocated 2421 MiB
00:35:50.105 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=8, keys=308, total=3777)] avg 496.92ms, allocated 14291 MiB
00:35:50.682 [main] INFO accord.txn.DepsTest - Benchmark[name=merge(size=2, keys=185, total=614)] avg 3.62ms, allocated 236 MiB
00:35:52.802 [main] INFO accord.txn.DepsTest - Benchmark[name=builder forEach(size=2, keys=185, total=614)] avg 42.05ms, allocated 1829 MiB
```

This is algorithmically much simpler than the priority heap merge, but it has involved quite a bit of ugliness with factories for managing memory allocation while still using the general purpose `SortedArrays` implementations of the core procedures. I'll continue to see if there's a little more that can be done to reduce garbage, but this is probably close to good enough now.

It's also still not testing the situation it should perform best for (i.e. merging lots of very similar dips)

We do have to decide if we simply discard the algorithmically "better" approach, or perhaps include it with a note that it has not been reviewed, as it might be useful in future and it would be a shame to have to start from scratch. I'm not sure of a better way to park this kind of code. A branch will get lost, but perhaps that's fine.

dcapwell commented:
> Benchmark[name=merge

Is this with `with`? or a new version of `merge`?

> We do have to decide if we simply discard the algorithmically "better" approach

I am a huge fan of "less is more", so if we can remove code and be in the same position I am all for it.  

> or perhaps include it with a note that it has not been reviewed

these things have a way to coming back to bite in the future, so if we keep it needs to go through review + testing.  

> I'm not sure of a better way to park this kind of code. A branch will get lost, but perhaps that's fine.

There is branch and gist... but the ability to "discovery" I find is always a pain...  If I need to find code I discard 6m path I have no idea how to find...

belliottsmith commented:
A new version of `merge` that calls `linearUnion` repeatedly, which is just a version of `with` that operates on arrays (so we can read and write from buffers). So both `with` and `merge` are now backed by the same underlying method.

This means a bit of ugliness for orchestrating buffers, and I have deliberately kept the actual caching of buffers really dumb/bad for now. It could probably be made a little bit nicer, but it is perhaps adequate.

dcapwell commented on accord-core/src/main/java/accord/local/Command.java:
```diff
@@ -224,22 +224,19 @@ public boolean commit(Txn txn, Key homeKey, Key progressKey, Timestamp executeAt
                     // we don't know when these dependencies will execute, and cannot execute until we do
-                    waitingOnCommit.put(id, command);
                     command.addListener(this);
+                    waitingOnCommit.put(txnId, command);
```
is there a reason to move this after adding the listener?  In both cases they seem to update a map but shouldn't depend on the other?

belliottsmith replied:
Probably just that it makes a kind of sense to ensure the listener is registered before we block execution on its membership of the collection, so we ensure the collection is maintained (though if there are no errors the two should be equivalent)

dcapwell commented on accord-core/src/main/java/accord/messages/PreAccept.java:
```diff
@@ -147,17 +148,23 @@ public String toString()
     }
 
-    static Dependencies calculateDeps(CommandStore commandStore, TxnId txnId, Txn txn, Timestamp executeAt)
+//    static Dependencies calculateDeps(CommandStore commandStore, TxnId txnId, Txn txn, Timestamp executeAt)
```
in my review branch I drop this commented out code.

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRange.java:
```diff
@@ -163,14 +163,22 @@ public String toString()
+     * Returns a negative integer, zero, or a positive integer as the provided key is greater than, contained by,
+     * or less than this range.
+     */
+    public abstract int compareTo(K key);
```
`@Override`

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+        Arrays.sort(this.keys);
+    }
+
+    public Keys(Key[] keys)
```
in my review branch I made this private so I could remove the `sort`, see adb09e7f2ce1084d3fe448e1c03e3312313a957c

For now I only allowed internal methods to use it to limit abuse, external classes go through `of`

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+        return keys.length;
+    }
+
+    public Keys select(int[] indexes)
```
no code touches this

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+        return wrap(SortedArrays.sliceWithOverlaps(keys, ranges.ranges, Key[]::new, (k, r) -> -r.compareTo(k), KeyRange::compareTo));
+    }
+
+    public int search(int lowerBound, int upperBound, Object key, Comparator<Object> comparator)
```
tried switching to `AsymmetricComparator` but the fact we are dealing with `Key` and `KeyRange<K extends Key>` requires `K`, hit a type mismatch =(

Would need `Keys` to be `Keys<K extends Key>` to resolve

belliottsmith replied:
I've brought forward the later patch's removal of the generic type parameter to this patch, and made use of `AsymmetricComparator`

dcapwell reviewed (commented):
will finish reviewing of `Keys` later, pushing updated results.

Finished `Deps` (without `merge`) and `SortedArray` so was able to get through a lot more today as I could start to work on their callers.  Now that those are done I might be able to complete this by Friday!  Will try hard to!

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRanges.java:
```diff
@@ -0,0 +1,395 @@
+        return rangeIndexForKey(key) >= 0;
+    }
+
+    public boolean containsAll(Keys keys)
```
no code uses,  this is also the only caller of `accord.primitives.Keys#rangeFoldl` so if we remove then can also remove that.

belliottsmith replied:
This is used in the follow-up patch, and was brought forward to reduce the review burden for that patch.

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+        return initialValue;
+    }
+
+    public boolean intersects(KeyRanges ranges, Keys keys)
```
not used, if we get rid of then can also drop `foldlIntersect(KeyRanges rs, Keys intersect`

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+     * and NOT the provided set.
+     */
+    @Inline
+    public long foldlDifference(Keys subtract, IndexedFoldToLong<Key> fold, long param, long initialValue, long terminalValue)
```
not used

belliottsmith replied:
It is used in the follow-up patch

dcapwell commented on accord-core/src/main/java/accord/primitives/Keys.java:
```diff
@@ -0,0 +1,449 @@
+        return as == null ? bs : bs == null ? as : as.union(bs);
+    }
+
+    public static Keys intersect(Keys as, Keys bs)
```
not used

dcapwell commented on accord-core/src/main/java/accord/utils/InlineHeap.java:
```diff
@@ -0,0 +1,317 @@
+ *
+ * }</pre>
+ */
+public class InlineHeap
```
not reviewing as the new `Deps.merge` doesn't need

dcapwell reviewed (commented):
ok done for the day, just have SimpleProgressLog.java and KeyRanges.java lest (plus all the tests), so close!

belliottsmith commented:
Ok, I think I have addressed all of the review feedback

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRanges.java:
```diff
@@ -0,0 +1,457 @@
+    {
+        for (int i = 1 ; i < ranges.length ; ++i)
+        {
+            if (ranges[i - 1].end().compareTo(ranges[i].start()) > 0)
```
```
[0, 10],
[10, 20]
```

when both `[I - 1].end` and `[I].start` are inclusive then there is an overlap

belliottsmith replied:
We currently _should_ *forbid* both start and end from being inclusive, as this simplified Blake's patch that introduced `KeyRanges`, so this isn't a problem. I did toy with removing this restriction, but it really does make a lot of stuff more complicated.

belliottsmith replied:
Hmm, I can't see where this is any longer enforced, except by the class hierarchy. It was previously in the `KeyRange` constructor I'm fairly sure, and it still should be.

dcapwell commented on accord-core/src/test/java/accord/utils/SortedArraysTest.java:
```diff
@@ -0,0 +1,335 @@
+class SortedArraysTest
+{
+//    @Test
+//    public void testRemapper()
```
adding this test back 521431ca9150816e70c42558b70bfc4006d65419

dcapwell commented on accord-core/src/main/java/accord/utils/SortedArrays.java:
```diff
@@ -0,0 +1,628 @@
+            }
+            result[i++] = j++;
+        }
+        Arrays.fill(result, i, srcLength, -1);
```
when target is the superset of src, src can not have a element not found in target else we proved target isn't a superset.  In the previous logic when superset=true I rejected, but with this rewrite we allow it again... now that the doc and name are clear only supersets are allowed, shouldn't we fail rather than put `-1`?

dcapwell replied:
Removed -1 in 4e2c60c219f34e261cea1abe256775f67bb205c2

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,1095 @@
+        return new Builder(keys);
+    }
+
+    private static class LinearMerger extends ArrayBuffers.SavingManager implements DepsConstructor<Object>
```
I added a patch to let me disable the cache logic and ran benchmarks with and without buffers... what I see is right now we don't get a win... so I feel we should drop the whole ArrayBuffers class, the logic in LinearMerge seems enough to make sure we can reuse between `Deps`

```
Benchmark[name=merge w/cache(size=15, keys=296, total=6994)] avg 487.11ms, allocated 6028 MiB
Benchmark[name=merge wo/cache(size=15, keys=296, total=6994)] avg 482.43ms, allocated 6028 MiB

Benchmark[name=merge w/cache(size=13, keys=355, total=5447)] avg 365.89ms, allocated 4530 MiB
Benchmark[name=merge wo/cache(size=13, keys=355, total=5447)] avg 364ms, allocated 4530 MiB

Benchmark[name=merge w/cache(size=5, keys=184, total=2435)] avg 45.33ms, allocated 1384 MiB
Benchmark[name=merge wo/cache(size=5, keys=184, total=2435)] avg 45.23ms, allocated 1384 MiB

Benchmark[name=merge w/cache(size=14, keys=335, total=5918)] avg 317.15ms, allocated 4006 MiB
Benchmark[name=merge wo/cache(size=14, keys=335, total=5918)] avg 323.47ms, allocated 4006 MiB

Benchmark[name=merge w/cache(size=18, keys=362, total=9583)] avg 679.28ms, allocated 12133 MiB
Benchmark[name=merge wo/cache(size=18, keys=362, total=9583)] avg 668.39ms, allocated 12692 MiB

Benchmark[name=merge w/cache(size=5, keys=237, total=3162)] avg 91.93ms, allocated 1731 MiB
Benchmark[name=merge wo/cache(size=5, keys=237, total=3162)] avg 83.33ms, allocated 1731 MiB

Benchmark[name=merge w/cache(size=3, keys=149, total=1662)] avg 20.64ms, allocated 807 MiB
Benchmark[name=merge wo/cache(size=3, keys=149, total=1662)] avg 20.67ms, allocated 807 MiB
```

sha c3dd834c008c1a98e71733a321c5e84ba9009269

belliottsmith replied:
As discussed offline, an updated patch using the latest caching logic and a similar split or enabled/disabled shows 50-90% memory reduction using caching.

dcapwell reviewed (commented):
pushed updated to my review branch and posting comments early.  What I see is right now that ArrayBuffers logic adds a lot of complexity, but doesn't have a clear win, so I feel this patch would be better dropping it all together.

belliottsmith commented:
I've pushed some further changes to improve javadoc, and in the process noticed that the `Deps.Builder` is unnecessarily costly for most use cases - since we only use it when building from `TxnId` that are visited in ascending order we can build into a single `int[]` and simply flip it.

I have chosen not to implement caching of these for now, but it's a trivial TODO for later. We could consider removing the `unorderedBuilder` entirely, as it's only needed in `DepsTest.generate` and `accord.maelstrom.Json` - both of which could readily manage without.

belliottsmith commented:
Ok, last round of changes I think. I've finished up the `OrderedBuilder` by using the same caching as merging, and have ported remaining use cases to the OrderedBuilder (besides the ones you introduced, which I will leave untouched until we're ready to merge)

I also introduced a merge built upon this builder, which produces a lot less garbage than the other builder approach but is slightly slower than the other builder approach (and *much* slower than the current merge approach). Once review is complete and we're satisfied with the current merge I will remove this little addition.

dcapwell commented on accord-core/src/main/java/accord/messages/BeginRecovery.java:
```diff
@@ -289,24 +273,44 @@ public String toString()
+                // committed txns with an earlier txnid and have our txnid as a dependency
+                builder.nextKey(key);
+                forKey.uncommitted.headMap(txnId, false).forEach((ts, command) -> {
+                    if (command.is(Accepted) && command.savedDeps().contains(txnId) && command.executeAt().compareTo(txnId) > 0)
```
this was `!c.savedDeps().contains(txnId)` before, now its searching explicitly for commands that depend on the TxnId?  Trying to understand this difference... is there a test I can look at that hits this subtle difference?

belliottsmith replied:
This is a good spot, and the only test that can easily elicit this is the burn test. It would be a difficult test to write to elicit this directly, but perhaps we should consider how to do so (or else have a tightly coupled test for the response given some simple state)

dcapwell commented on accord-core/src/main/java/accord/messages/BeginRecovery.java:
```diff
@@ -76,19 +76,13 @@ public void process(Node node, Id replyToNode, ReplyContext replyContext)
                 // accepted txns with an earlier txnid that don't have our txnid as a dependency
-                earlierAcceptedNoWitness = uncommittedStartedBefore(instance, txnId, txn.keys)
-                                              .filter(c -> c.is(Accepted)
-                                                           && !c.savedDeps().contains(txnId)
```
trying to understand this difference... we now do `command.savedDeps().contains(txnId)`... is there a test I can look at that hits this subtle difference?

dcapwell commented on accord-core/src/main/java/accord/primitives/Deps.java:
```diff
@@ -0,0 +1,1394 @@
+    }
+
+    // TODO: this is slower, so don't commit it - but keeping around for review purposes
+    public static <T> Deps mergeWithBuilder(Keys keys, List<T> merge, Function<T, Deps> getter)
```
cool to drop now.

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRanges.java:
```diff
@@ -0,0 +1,456 @@
+
+    public int rangeIndexForKey(int lowerBound, int upperBound, Key key)
+    {
+        return Arrays.binarySearch(ranges, lowerBound, upperBound, key,
```
now that generics are removed, can replace with `return SortedArrays.binarySearch(ranges, lowerBound, upperBound, key, (k, r) -> r.compareKey(k), SortedArrays.Search.FAST);`

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRanges.java:
```diff
@@ -0,0 +1,456 @@
+        KeyRange[] selection = new KeyRange[indexes.length];
+        for (int i=0; i<indexes.length; i++)
+            selection[i] = ranges[indexes[i]];
+        return new KeyRanges(selection);
```
nothing validates the inputs are sorted, so should use `ofSortedAndDeoverlapped`

dcapwell commented on accord-core/src/main/java/accord/primitives/KeyRanges.java:
```diff
@@ -0,0 +1,456 @@
+                trg[trgPosition + count++] = maybeUpdateEnd(prev, end);
+                prev = next;
+                end = next.end();
+            }
```
could replace with

```
if (!end.equals(next.start()))
            {
                trg[trgPosition + count++] = maybeUpdateEnd(prev, end);
                prev = next;
            }
            end = next.end();
```

dcapwell commented on accord-core/src/main/java/accord/impl/SimpleProgressLog.java:
```diff
@@ -375,6 +377,8 @@ public String toString()
         {
             this.blocking = blocking;
+            if (blocking.txn() != null && !blocking.txn().keys.containsAll(someKeys))
+                throw new IllegalStateException();
```
would be good to produce a meaningful error, so if this ever happens we have enough to debug

belliottsmith replied:
I think this will only come up in BurnTests where it's easily reproducible, and I think this line is being removed anyway in the next patch. But I'll add some for now anyway.

dcapwell commented on accord-core/src/main/java/accord/utils/ArrayBuffers.java:
```diff
@@ -0,0 +1,548 @@
+
+        public int lengthOfLast(T[] buffer)
+        {
+            return length;
```
this can return `-1` which means we called without a call to `complete`, so rather than returning `-1` (which no caller handles) we should fail

dcapwell reviewed (commented):
finally finished!  Take a look at the changes in review/minimal-deps, most are nits but the changes to ArrayBuffers I think is to make sure callers don't get unexpected values

belliottsmith commented:
Ok, I think we're ready to commit

belliottsmith committed:
Efficient Dependencies (CASSANDRA-17111)

Transaction dependencies must include only the TxnId of the dependent transactions, and the key they conflict on

patch by Benedict; reviewed by David Capwell for CASSANDRA-17111
